Optax QSP Iterative Solver

cvqmm opened the pull request:
**Context:** Use Optax optimizers instead of Scipy's

**Description of the Change:**

**Benefits:** JAX friendly & whole optimization loop can be jit-compiled

**Possible Drawbacks:**

**Related GitHub Issues:**
[sc-110370] [sc-110735]

Commit:
[no ci] bump nightly version

Commit:
[no ci] bump nightly version

Commit:
[no ci] bump nightly version

JerryChen97 commented on pennylane/_version.py:
We can revert this change and then when all the other conflicts are resolved merge master into this branch then files like `_version` should be automatically updated

JerryChen97 reviewed (changes requested):
Thanks a lot for the work! Some workflow changes are needed, and then let's merge master into this branch and  I'll approve the CI to run more auto checks

cvqmm committed:
optax based iterative optimizer

cvqmm committed:
update tests

Commit:
[no ci] bump nightly version

cvqmm committed:
thin wrapper around jit to fix tests

Commit:
[no ci] bump nightly version

cvqmm committed:
minor

cvqmm committed:
fix tests

cvqmm committed:
formatting

cvqmm committed:
Merge branch 'master' into optax_qsp_iter_solver

cvqmm committed:
marking jax tests

JerryChen97 committed:
Merge branch 'master' into optax_qsp_iter_solver

JerryChen97 commented:
> > Thanks a lot for the work! Some workflow changes are needed, and then let's merge master into this branch and I'll approve the CI to run more auto checks
> 
> I think all the comments can be resolved now (?). The failing tests need Optax to be passed as an additional_python_packages. @JerryChen97 Can you plz look into it? Thx :)

I'll discuss within team the choice of either install optax for jax CI or put the tests into external.

Commit:
[no ci] bump nightly version

cvqmm committed:
marking optax-requiring tests as external

cvqmm committed:
Merge branch 'master' into optax_qsp_iter_solver

JerryChen97 committed:
Merge branch 'master' into optax_qsp_iter_solver

JerryChen97 commented on pennylane/templates/subroutines/qsvt.py:
```diff
@@ -929,97 +936,92 @@ def _grid_pts(degree, interface):
+    try:
+        import optax
+    except ModuleNotFoundError:
+        raise ModuleNotFoundError(f"Optax is required but not installed")
```
```suggestion
    except ModuleNotFoundError:  # pragma: no cover
        raise ModuleNotFoundError(f"Optax is required but not installed")  # pragma: no cover
```

cvqmm replied:
should be covered now https://github.com/PennyLaneAI/pennylane/blob/4f2a40ae9b71f240ad11f27350d677cf1dd64912/tests/templates/subroutines/test_qsvt.py#L1047

JerryChen97 reviewed (commented):
Thanks a lot for another round of changes! Finally all the tests passed and now we have codecov's report. They look legit and it will be fantastic if we can also resolve them. Most are just some module import excepts totally skippable, but the missing fallback branch should be tested.

cvqmm committed:
make format

Commit:
[no ci] bump nightly version

cvqmm committed:
test thin jit wrapper

cvqmm committed:
"untestable" try/catch

Co-authored-by: Yushao Chen (Jerry) <chenys13@outlook.com>

cvqmm committed:
squeeze two try/catch blocks into one

cvqmm committed:
test fallout when jax &/| optax are not installed

Commit:
[no ci] bump nightly version

Commit:
[no ci] bump nightly version

cvqmm committed:
make format

JerryChen97 commented on tests/templates/subroutines/test_qsvt.py:
```diff
@@ -991,27 +999,59 @@ def test_qsp_on_poly_with_parity(self, polynomial_coeffs_in_cheby_basis):
+            #             for i in range(len(target_polynomial_coeffs))
+            #         ]
+            #     )
+            # )
```
do we still need them?

JerryChen97 commented on pennylane/templates/subroutines/qsvt.py:
```diff
@@ -929,97 +938,89 @@ def _grid_pts(degree, interface):
-        degree (int): degree of polynomial function
-        coeffs_target_func (tensor_like): coefficients of the polynomial function in ascending index order
+@jit_if_jax_available
+def obj_function(phi, x, y):
```
one or two sentences talking about what this function is doing?

Commit:
[no ci] bump nightly version

cvqmm committed:
remove commented obsolete code

cvqmm committed:
doc & silencing pylint unused imports

Co-authored-by: Yushao Chen (Jerry) <chenys13@outlook.com>

cvqmm committed:
silencing pylint unused import warnings

Co-authored-by: Yushao Chen (Jerry) <chenys13@outlook.com>

cvqmm committed:
Update pennylane/templates/subroutines/qsvt.py

Co-authored-by: Yushao Chen (Jerry) <chenys13@outlook.com>

cvqmm committed:
Update pennylane/templates/subroutines/qsvt.py

Co-authored-by: Yushao Chen (Jerry) <chenys13@outlook.com>

cvqmm committed:
replacing imports try/catch blocks with indicative variable

cvqmm committed:
minor

cvqmm committed:
clean up

cvqmm committed:
update jit wrapper

cvqmm commented on pennylane/templates/subroutines/qsvt.py:
like this?https://github.com/PennyLaneAI/pennylane/blob/7a6869c1e356d2a7539f1c8aa409d0d1828646d4/pennylane/templates/subroutines/qsvt.py#L43-L48

cvqmm committed:
docs

cvqmm committed:
make format

cvqmm committed:
Merge branch 'master' into optax_qsp_iter_solver

JerryChen97 committed:
Merge branch 'master' into optax_qsp_iter_solver

JerryChen97 committed:
Merge branch 'master' into optax_qsp_iter_solver

cvqmm committed:
remove superfluous else condition

Co-authored-by: Yushao Chen (Jerry) <chenys13@outlook.com>

JerryChen97 committed:
Merge branch 'master' into optax_qsp_iter_solver

KetpuntoG reviewed (commented):
Great work here! 😄
Here some minor comments

cvqmm committed:
fix typo

Co-authored-by: Guillermo Alonso-Linaje <65235481+KetpuntoG@users.noreply.github.com>

cvqmm commented on tests/templates/subroutines/test_qsvt.py:
```diff
@@ -974,44 +991,60 @@ class TestIterativeSolver:
+        print(
+            _qsp_iterate_broadcast(phis, x_point, "jax")
+            - _poly_func(coeffs=target_polynomial_coeffs, x=x_point)
+        )
```
the test is already done in the assert statement in the line above.. this is just to display the error.. can be removed

KetpuntoG committed:
Merge branch 'master' into optax_qsp_iter_solver

KetpuntoG commented:
Nice work! After the tests pass (it seems just a "black" issue), I'll approve the PR 😄

cvqmm committed:
fix typo

cvqmm committed:
remove print

cvqmm committed:
remove commented code

cvqmm committed:
minor

cvqmm committed:
minor

DSGuala committed:
Merge branch 'master' into optax_qsp_iter_solver

KetpuntoG commented on pennylane/templates/subroutines/qsvt.py:
```diff
@@ -929,97 +940,97 @@ def _grid_pts(degree, interface):
     )
 
 
-def _qsp_optimization(degree, coeffs_target_func, interface=None):
```
What I see is that algorithmically speaking they are equivalent. The only difference is that they've switched from using SciPy to OptaX to make it jax/jit compatible (which is what we aim for). Apart from that, there is (on my end) no benchmark done in the workflow

JerryChen97 commented on pennylane/templates/subroutines/qsvt.py:
```diff
@@ -41,6 +40,29 @@
+try:
+    import jax
+except ImportError:
+    is_jax_available = False
```
```suggestion
except ImportError:  # pragma: no cover
    is_jax_available = False  # pragma: no cover
```

JerryChen97 commented on pennylane/templates/subroutines/qsvt.py:
```diff
@@ -929,97 +940,97 @@ def _grid_pts(degree, interface):
-def _compute_qsp_angles_iteratively(poly):
-    """Calculates the angles given a polynomial in canonical base
+    if not is_optax_available:
+        raise ModuleNotFoundError("optax is required!")
```
```suggestion
        raise ModuleNotFoundError("optax is required!")  # pragma: no cover
```

cvqmm committed:
minor

Co-authored-by: Yushao Chen (Jerry) <chenys13@outlook.com>

cvqmm committed:
minor

Co-authored-by: Yushao Chen (Jerry) <chenys13@outlook.com>

JerryChen97 committed:
Merge branch 'master' into optax_qsp_iter_solver

cvqmm committed:
remove redundant jax imports

cvqmm committed:
make format

JerryChen97 committed:
Merge branch 'master' into optax_qsp_iter_solver

JerryChen97 committed:
Merge branch 'master' into optax_qsp_iter_solver

cvqmm committed:
docstring of optax_opt

Co-authored-by: Yushao Chen (Jerry) <chenys13@outlook.com>

cvqmm committed:
replace lambda function

Co-authored-by: Yushao Chen (Jerry) <chenys13@outlook.com>

JerryChen97 committed:
Merge branch 'master' into optax_qsp_iter_solver

cvqmm committed:
make format

JerryChen97 committed:
Merge branch 'master' into optax_qsp_iter_solver

JerryChen97 committed:
Merge branch 'master' into optax_qsp_iter_solver

JerryChen97 reviewed (approved):
Thanks a lot for the patience!

KetpuntoG committed:
Merge branch 'master' into optax_qsp_iter_solver

KetpuntoG reviewed (approved):
Great work! 🤩🚀

maliasadi commented on pennylane/templates/subroutines/qsvt.py:
```diff
@@ -41,6 +40,29 @@
+try:
+    import optax
+except ImportError:
+    is_otpax_available = False
```
To avoid global variables, these could be encapsulated into the availability checks in a function or a dependencies dataclass instead of mutating module-level state.

``` python
def check_dependencies():
    try:
        import jax
        jax_available = True
    except ImportError:  # pragma: no cover
        jax_available = False

    try:
        import optax
        optax_available = True
    except ImportError:  # pragma: no cover
        optax_available = False

    return jax_available, optax_available
```

Another way to tackle this here would be by using `importlib.util.find_spec`:

``` python
def is_available(module_name) -> bool:
    return importlib.util.find_spec(module_name) is not None
```

JerryChen97 replied:
I tried wrapping `jax` and `optax` availability in a function closure and there would be quite some more local imports in this PR to support the feature it tries to add.
See https://github.com/PennyLaneAI/pennylane/tree/local-jax-import for details. And would like to hear what @cvqmm thinks about this.

maliasadi committed:
Merge branch 'master' into optax_qsp_iter_solver

JerryChen97 committed:
Remove rc0 from version number

cvqmm committed:
suggestion from reviews

cvqmm committed:
minor

JerryChen97 committed:
`CompilePipeline.extend` to handle `list`s and `CompilePipeline` objects (#8820)

**Context:**

**Description of the Change:**
- `extend` acts as a typical `extend` method a Python user would expect
for any list.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

---------

Co-authored-by: Astral Cai <astral.cai@xanadu.ai>

ringo-but-quantum committed:
[no ci] bump nightly version

cvqmm committed:
Merge branch 'v0.44.0-rc0' into optax_qsp_iter_solver

astralcai commented on pennylane/templates/subroutines/qsvt.py:
```diff
@@ -821,22 +859,16 @@ def _cheby_pol(x, degree):
-    return sum(coeffs[i] * _cheby_pol(x, degree=2 * i + parity) for i in ind)
+    return jax.numpy.sum(
+        coeffs @ jax.vmap(_cheby_pol, in_axes=(None, 0))(x, np.arange(coeffs.shape[0]))
+    )
```
Is this function ever called in a context where `jax` isn't available?

albi3ro commented on pennylane/templates/subroutines/qsvt.py:
```diff
@@ -929,97 +955,101 @@ def _grid_pts(degree, interface):
+    r"""Algorithm 1 in https://arxiv.org/pdf/2002.11649 produces the angle parameters by minimizing the distance between the target and qsp polynomial over the grid"""
 
-        obj_func = math.dot(obj_func, obj_func)
+    jax.config.update("jax_enable_x64", True)
```
I would instead prefer to warn or error if x64 mode is not enabled.  We should not be changing the global configuration under the hood.
```suggestion
```

albi3ro reviewed (dismissed):
Only a scan, but I will block on updating the x64 mode sneekily in the source code.

cvqmm committed:
remove commented code

Co-authored-by: Ali Asadi <10773383+maliasadi@users.noreply.github.com>

cvqmm committed:
make foramt & remove commented code

cvqmm committed:
Update pennylane/templates/subroutines/qsvt.py

Co-authored-by: Christina Lee <chrissie.c.l@gmail.com>

JerryChen97 committed:
Merge branch 'master' into optax_qsp_iter_solver

cvqmm commented:
@JerryChen97 is there any unresolved comments that need to be addressed still?

JerryChen97 committed:
Merge branch 'master' into optax_qsp_iter_solver

JerryChen97 commented:
> @JerryChen97 is there any unresolved comments that need to be addressed still?

@cvqmm Not from my end! I'll re-request other reviewers for further confirmation.

JerryChen97 reviewed (changes requested):
We should also resolve the regression issue! The non-jax routine in the past should also be available instead of having QSVT only runnable with jax and optax

JerryChen97 commented on pennylane/templates/subroutines/qsvt.py:
```diff
@@ -821,22 +847,16 @@ def _cheby_pol(x, degree):
 
-    ind = math.arange(len(coeffs))
-    return sum(coeffs[i] * _cheby_pol(x, degree=2 * i + parity) for i in ind)
+    return jax.numpy.sum(
```
yeah pennylane does not require jax as dependency, and if now `QSVT` requires jax then it's breaking the previous functionality we provided. Better solution is if the user has jax and optax we use the new routine, otherwise fallback to previous

JerryChen97 committed:
Merge branch 'master' into optax_qsp_iter_solver

JerryChen97 committed:
Bring back the scipy path for qsvt solver:
1. Users with no jax or optax will use scipy
2. Docstrings have been updated to reflect the change, and the test cases have been modified to test both optax and scipy solvers.

JerryChen97 committed:
add tests for both scipy and optax

JerryChen97 committed:
Merge branch 'master' into feature/optax_qsp_solver

JerryChen97 committed:
fix test_qsvt.py: scipy not external; optax external

JerryChen97 committed:
Merge branch 'master' into optax_qsp_iter_solver

JerryChen97 committed:
Major refactor: separate file for test qsvt optax

JerryChen97 committed:
import or skip

JerryChen97 committed:
fix test qsvt optax

JerryChen97 committed:
code cov

JerryChen97 committed:
debug cov issue

JerryChen97 committed:
mitigate codecov bug

JerryChen97 committed:
revert unnecessary error msg alteration

JerryChen97 committed:
clean some unnecessary docstr as suggested separately from @DSGuala

JerryChen97 committed:
merge two veresions of cheby_pol

JerryChen97 committed:
merge grid_pts

JerryChen97 committed:
Merge branch 'master' into optax_qsp_iter_solver

JerryChen97 committed:
Merge branch 'master' into feature/optax_qsp_solver

JerryChen97 committed:
Merge branch 'optax_qsp_iter_solver' into feature/optax_qsp_solver

JerryChen97 committed:
Improve the docstring by highlighting the correct usage of jit stuff

DSGuala committed:
iterative_optax -> iterative-optax

DSGuala committed:
update docstrings

JerryChen97 committed:
remove placeholders and make format

cvqmm committed:
Merge pull request #3 from PennyLaneAI/feature/optax_qsp_solver

Backward compatible optax qsp solver

JerryChen97 committed:
Merge branch 'master' into optax_qsp_iter_solver

cvqmm committed:
minor

cvqmm committed:
remove duplicate code

cvqmm committed:
refac

JerryChen97 committed:
disable pylint

cvqmm committed:
black

cvqmm committed:
format

cvqmm committed:
make format
